### PR TITLE
Dont remove volumes when saveVolumeData fails

### PR DESCRIPTION
**Pull request by huffmanca**
**What type of PR is this?**

/kind bug

**What this PR does / why we need it**:
This is a continuation of https://github.com/kubernetes/kubernetes/pull/89464 . It incorporates that author's commit while also adjusting the location of the `defer()` function to correctly remove.

**Which issue(s) this PR fixes**:
Fixes # #89281

**Special notes for your reviewer**:

**Does this PR introduce a user-facing change?**:
```release-note
All data is no longer automatically deleted when a failure is detected during creation of the volume data file on a CSI volume. Now only the data file and volume path is removed.
```

**Additional documentation e.g., KEPs (Kubernetes Enhancement Proposals), usage docs, etc.**:

### CODE REVIEW

File: pkg/volume/csi/csi_plugin.go [hunk 1]
```diff
@@ -435,11 +436,23 @@ func (p *csiPlugin) NewMounter(
 	attachID := getAttachmentName(volumeHandle, driverName, node)
 	volData[volDataKey.attachmentID] = attachID
 
-	if err := saveVolumeData(dataDir, volDataFileName, volData); err != nil {
-		if removeErr := os.RemoveAll(dataDir); removeErr != nil {
-			klog.Error(log("failed to remove dir after error [%s]: %v", dataDir, removeErr))
+	err = saveVolumeData(dataDir, volDataFileName, volData)
```
Comment (gnufied) [inline pkg/volume/csi/csi_plugin.go:439]: should we do same thing in this file for other places where `saveVolumeData` is being called followed up by `os.RemoveAll`? This applies to `NewBlockVolumeMapper` call.

**Comment by gnufied** [comment]
@huffmanca have we tried unit testing this stuff? it might be difficult but worth giving it a shot I think.

File: pkg/volume/csi/csi_attacher.go [hunk 2]
```diff
@@ -296,25 +296,26 @@ func (c *csiAttacher) MountDevice(spec *volume.Spec, devicePath string, deviceMo
...
+
+	err = saveVolumeData(dataDir, volDataFileName, data)
 	defer func() {
 		// Only if there was an error and volume operation was considered
 		// finished, we should remove the directory.
 		if err != nil && volumetypes.IsOperationFinishedError(err) {
 			// clean up metadata
 			klog.Errorf(log("attacher.MountDevice failed: %v", err))
-			if err := removeMountDir(c.plugin, deviceMountPath); err != nil {
+			if err := os.RemoveAll(dataDir); err != nil {
```
Comment (gnufied) [inline pkg/volume/csi/csi_attacher.go:307]: Isn't using `os.RemoveAll` opposite of what we set out to fix? This means if - saving the `volDataFileName` fails then we will still end up calling `rm -rf` on entire data directory?

**Comment by gnufied** [comment]
@huffmanca can you add a release-note plz? I think this is important enough of a change that should be covered in release notes.

File: pkg/volume/csi/csi_plugin.go (hunk 1 above)
Comment (msau42) [reply pkg/volume/csi/csi_plugin.go:460]: Is this directory same or different than the one used in MountDevice?

Comment (huffmanca) [reply pkg/volume/csi/csi_plugin.go:460]: As best I can tell from testing with the hostpath driver, these appear to be different directories:

* csi_plugin: created path successfully [/var/lib/kubelet/pods/103bb9c7-0a9f-4958-bb6a-8d05f64420f5/volumes/kubernetes.io~csi/pvc-21228cc8-5669-4c0d-bbdc-d7b4e19dfb79]
* csi_attacher: created target path successfully [/var/lib/kubelet/plugins/kubernetes.io/csi/pv/pvc-22c63cbf-4e2c-4ca6-b269-d5e44a1f3718/globalmount]

**Pull request merged**